### ISSUE TITLE
Rotating Fonts

### ISSUE THREAD

**Issue by 12AT7** [Blatko1/wgpu-text#15]
Wgpu-text is working pretty well for me for standard use cases.  I next tried to rotate my text, so that it reads not left-to-right but top-to-bottom; I did this to try to cram in more x-axis labels on a numerical plot that does not have enough room.

I messed around quite a bit with `update_matrix(...)` and it also works.  However, the new transform seems to affect both the rendering of the font itself in addition to my other positioning commands to get the annotation in the right spot.  My glyphs are also being distorted by my positioning and scaling to the clip coordinates; I want to move and rotate the glyph, but not actually change the glyph itself.

I think this is because I need to rotate the glyphs independently of the positioning transform.   Like maybe I need a function like `Section::with_orientation()` that is used similarly to `Section::with_screen_location()` except that it transforms the glyph before the placement.

Is this a good strategy?  Or am I making this too hard (hopefully this is the case)?   I also looked at the underlying `ab_glyph` library to see if it had any useful features that could help, but as far as I can tell the whole font rendering stack just assumes that the glyphs should be rendered to read horizontally.

Am I missing a simple way to render vertically oriented fonts here?

**Comment by Blatko1** [comment]
In this crate's inner workings, all it does is process all queued text (done through the `glyph-brush` crate), and after retrieving all text vertices, it just uploads them to the vertex buffer (used for drawing). Vertices are returned in a single large `Vec`, so it's impossible to recognize a single `glyph` or a `Section`. Because all glyph crunching and glyph calculations happen in the `glyph-brush` crate, this crate cannot affect any of it.

Sadly, as you stated, there are no existing functions for applying rotations to a `Section`. Devs at [glyph-brush](https://github.com/alexheretic/glyph-brush/issues) would need to implement such functions at their end for `rotating` to work in this crate. The `update_matrix(...)` also doesn't provide much help since it affects all rendered text. Also, I think you were correct when you said that *the whole font rendering stack just assumes that the glyphs should be rendered to read horizontally.*

One way I can think of so rotations could be implemented is that each `Section` would need to accept a custom matrix which the `Section` would apply to its already calculated vertices. Most of it would be performed in the `glyph-brush` crate. Indeed, there are more effective ways this all could be done.

Therefore, there is no *standard* way to rotate a `Section`, but there are some out-of-the-box ways. For example, you can create a new `TextBrush` for each rotated section and apply custom matrices through the `update_matrix(...)` function. If you need only vertically oriented glyphs, you can use a unique font with only vertically oriented letters and another typically oriented, then swap between them. 

You can also try an alternative text rendering method which utilizes *signed distance fields* or, if you need higher quality, *multi-channel signed distance fields*. You can read more on it [here](https://github.com/Blatko1/awesome-msdf).

I'll leave this issue open in case you have some other questions or some news pop up.

**Comment by 12AT7** [comment]
Yeah, I had the same idea about using two brushes and I tried it, and it works.   It's not pretty because for positioning, all the x's and y's and layouts are swapped.  Also I had problems with getting mirrored text (unreadable).

But it _does_ work in the end, and looks nice in the rendering.   I'll just drop a comment in my code about this.  I won't be looking at that much after all once it is working.

Thanks for the thoughtful note!

**Comment by PPakalns** [comment]
I have similar issue and I am using patched wgpu_text to resolve it.

In my case y axis for world is in opposite direction of assumed y axis direction in wgpu_text. (I need to flip y axis of the text)

I have tried to resolve it by modifying matrix or using negatvie position coordinates, but it looks like to flip text (y axis) I need additional information about texture box size etc.
Maybe I am doing something wrong.

But to resolve the issue I just modified shader to render text in needed direction.

https://github.com/Blatko1/wgpu-text/commit/00194e8643bc7cd9ac9dd60020c6f2b51434100a

**Comment by Blatko1** [comment]
> I have similar issue and I am using patched wgpu_text to resolve it.
> 
> In my case y axis for world is in opposite direction of assumed y axis direction in wgpu_text. (I need to flip y axis of the text)
> 
> I have tried to resolve it by modifying matrix or using negative position coordinates, but it looks like to flip text (y axis) I need additional information about texture box size etc. Maybe I am doing something wrong.
> 
> But to resolve the issue I just modified shader to render text in needed direction.
> 
> [00194e8](https://github.com/Blatko1/wgpu-text/commit/00194e8643bc7cd9ac9dd60020c6f2b51434100a)

@PPakalns Please transfer your question to the **Discussion** page or create a new discussion and link it here, as I might have a solution for your problem. I will provide help there.
